python/ray/data/tests/test_execution_optimizer.py

```diff
@@ -374,6 +380,74 @@ def _check_repartition_usage_and_stats(ds):
     _check_repartition_usage_and_stats(ds)
 
 
+@pytest.mark.parametrize("preserve_order", (True, False))
+def test_union_operator(ray_start_regular_shared, enable_optimizer, preserve_order):
+    planner = Planner()
+    read_parquet_op = Read(ParquetDatasource(), [])
+    read_range_op = Read(RangeDatasource(), [])
+    read_json_op = Read(JSONDatasource(), [])
+    union_op = Union(
+        read_parquet_op,
+        read_range_op,
+        read_json_op,
+    )
+    plan = LogicalPlan(union_op)
+    physical_op = planner.plan(plan).dag
+
+    assert union_op.name == "Union"
+    assert isinstance(physical_op, UnionOperator)
+    assert len(physical_op.input_dependencies) == 3
+    for input_op in physical_op.input_dependencies:
+        assert isinstance(input_op, MapOperator)
```

Comment:
can you also add some testing code about `add_input`, `input_done`, `has_next`, `get_next`?
For example, when preserve_order = True, call add_input(0), has_next should return true. get_next then call add_input(1) then has_next should return false. then call `input_done(0)`, has_next should return true.